src/term/transform/extract_matches.rs

```diff
@@ -0,0 +1,215 @@
+use crate::term::{
+  check::type_check::{infer_arg_type, Type},
+  display::DisplayJoin,
+  Adt, Book, DefId, DefNames, Definition, Name, Pattern, Rule, Tag, Term,
+};
+use indexmap::IndexSet;
+use itertools::Itertools;
+use std::collections::{BTreeMap, HashMap, HashSet};
+
+impl Book {
+  /// Extracts any match terms into pattern matching functions.
+  pub fn extract_matches(&mut self) -> Result<(), String> {
+    let book = &mut MatchesBook::new(&mut self.def_names);
+
+    for (def_id, def) in &mut self.defs {
+      let def_name = book.def_names.name(def_id).cloned().unwrap();
+      for rule in def.rules.iter_mut() {
+        rule.body.extract_matches(&def_name, book, &mut 0).map_err(|err| err.to_string())?;
+      }
+    }
+
+    if !book.new_defs.is_empty() {
+      self.defs.append(&mut book.new_defs);
+      self.flatten_rules();
```

Comment:
this can be removed if flatten_rules are used after the `extract_matches` pass (as it is now in the order on the `encode_pattern_matching`). 

We can keep in case that order could change, but adding a post-condition documentation that flatten_rules should be used after this could be enough.

Reply:
Didn't even notice this line here. I think it's better to remove it and have all the passes be sort of complete in themselves and not call other passes, since that increases the complexity of the code.